File: rust/crates/infra/src/video_repo.rs
use sqlx::PgPool;

use gabon_shared::error::AppError;
use gabon_shared::traits::{MyVideoRow, VideoDetailRow, VideoListRow};

pub struct PgVideoRepo<'a> {
    pub pool: &'a PgPool,
}

impl gabon_shared::traits::VideoRepo for PgVideoRepo<'_> {
    async fn list_approved(
        &self,
        page: i64,
        page_size: i64,
        keyword: Option<&str>,
    ) -> Result<(Vec<VideoListRow>, i64), AppError> {
        let offset = (page - 1) * page_size;

        let (total, items) = if let Some(kw) = keyword {
            let pattern = format!("%{kw}%");
            let total: i64 = sqlx::query_scalar(
                "SELECT COUNT(*) FROM videos WHERE status = 4 AND deleted_at IS NULL AND title ILIKE $1",
            )
            .bind(&pattern)
            .fetch_one(self.pool)
            .await?;

            let items = sqlx::query_as::<_, VideoListRow>(
                r"SELECT id, customer_id, title, thumbnail_url, duration, like_count, total_clicks
                   FROM videos WHERE status = 4 AND deleted_at IS NULL AND title ILIKE $1
                   ORDER BY id DESC LIMIT $2 OFFSET $3",
            )
            .bind(&pattern)
            .bind(page_size)
            .bind(offset)
            .fetch_all(self.pool)
            .await?;

            (total, items)
        } else {
            let total: i64 = sqlx::query_scalar(
                "SELECT COUNT(*) FROM videos WHERE status = 4 AND deleted_at IS NULL",
            )
            .fetch_one(self.pool)
            .await?;

            let items = sqlx::query_as::<_, VideoListRow>(
                r"SELECT id, customer_id, title, thumbnail_url, duration, like_count, total_clicks
                   FROM videos WHERE status = 4 AND deleted_at IS NULL
                   ORDER BY id DESC LIMIT $1 OFFSET $2",
            )
            .bind(page_size)
            .bind(offset)
            .fetch_all(self.pool)
            .await?;

            (total, items)
        };

        Ok((items, total))
    }

    async fn list_featured(
        &self,
        page: i64,
        page_size: i64,
    ) -> Result<(Vec<VideoListRow>, i64), AppError> {
        let offset = (page - 1) * page_size;

        let total: i64 = sqlx::query_scalar(
            "SELECT COUNT(*) FROM videos WHERE status = 4 AND deleted_at IS NULL",
        )
        .fetch_one(self.pool)
        .await?;

        let items = sqlx::query_as::<_, VideoListRow>(
            r"SELECT id, customer_id, title, thumbnail_url, duration, like_count, total_clicks
               FROM videos WHERE status = 4 AND deleted_at IS NULL
               ORDER BY like_count DESC, id DESC LIMIT $1 OFFSET $2",
        )
        .bind(page_size)
        .bind(offset)
        .fetch_all(self.pool)
        .await?;

        Ok((items, total))
    }

    async fn list_my(&self, customer_id: i64) -> Result<Vec<MyVideoRow>, AppError> {
        let items = sqlx::query_as::<_, MyVideoRow>(
            r"SELECT id, title, thumbnail_url, file_url, duration, status,
                      like_count, total_clicks, valid_clicks
               FROM videos WHERE customer_id = $1 AND deleted_at IS NULL
               ORDER BY id DESC",
        )
        .bind(customer_id)
        .fetch_all(self.pool)
        .await?;

        Ok(items)
    }

    async fn like(&self, video_id: i64, customer_id: i64) -> Result<bool, AppError> {
        let result = sqlx::query(
            r"WITH inserted AS (
                   INSERT INTO video_likes (video_id, customer_id) VALUES ($1, $2)
                   ON CONFLICT DO NOTHING RETURNING video_id
               )
               UPDATE videos SET like_count = like_count + 1, updated_at = NOW()
               WHERE id = (SELECT video_id FROM inserted)",
        )
        .bind(video_id)
        .bind(customer_id)
        .execute(self.pool)
        .await?;
        Ok(result.rows_affected() > 0)
    }

    async fn unlike(&self, video_id: i64, customer_id: i64) -> Result<bool, AppError> {
        let result = sqlx::query(
            r"WITH deleted AS (
                   DELETE FROM video_likes WHERE video_id = $1 AND customer_id = $2
                   RETURNING video_id
               )
               UPDATE videos SET like_count = like_count - 1, updated_at = NOW()
               WHERE id = (SELECT video_id FROM deleted)",
        )
        .bind(video_id)
        .bind(customer_id)
        .execute(self.pool)
        .await?;
        Ok(result.rows_affected() > 0)
    }

    async fn record_play(
        &self,
        video_id: i64,
        customer_id: Option<i64>,
        play_type: i16,
    ) -> Result<i64, AppError> {
        let col = if play_type == 1 { "total_clicks" } else { "valid_clicks" };
        sqlx::query(&format!(
            "UPDATE videos SET {col} = {col} + 1, updated_at = NOW() WHERE id = $1 AND deleted_at IS NULL"
        ))
        .bind(video_id)
        .execute(self.pool)
        .await?;

        let id: i64 = sqlx::query_scalar(
            r"INSERT INTO video_play_records (video_id, customer_id, play_type, ip_address)
               VALUES ($1, $2, $3, NULL) RETURNING id",
        )
        .bind(video_id)
        .bind(customer_id)
        .bind(play_type)
        .fetch_one(self.pool)
        .await?;
        Ok(id)
    }

    async fn get_detail(&self, video_id: i64, viewer_id: Option<i64>) -> Result<Option<VideoDetailRow>, AppError> {
        let row = sqlx::query_as::<_, PgVideoDetailPartial>(
            r"SELECT v.id, v.title, v.description, v.file_url, v.thumbnail_url,
                      v.duration, v.like_count, v.total_clicks,
                      c.id AS author_id, c.name AS author_name,
                      c.avatar_url AS author_avatar_url, c.is_vip AS author_is_vip
               FROM videos v JOIN customers c ON v.customer_id = c.id
               WHERE v.id = $1 AND v.status = 4 AND v.deleted_at IS NULL",
        )
        .bind(video_id)
        .fetch_optional(self.pool)
        .await?;

        let Some(row) = row else { return Ok(None) };

        let (viewer_liked, viewer_followed) = if let Some(uid) = viewer_id {
            let liked: bool = sqlx::query_scalar(
                "SELECT EXISTS(SELECT 1 FROM video_likes WHERE video_id = $1 AND customer_id = $2)",
            )
            .bind(video_id)
            .bind(uid)
            .fetch_one(self.pool)
            .await?;
            let followed: bool = sqlx::query_scalar(
                "SELECT EXISTS(SELECT 1 FROM user_follows WHERE follower_id = $1 AND followed_id = $2)",
            )
            .bind(uid)
            .bind(row.author_id)
            .fetch_one(self.pool)
            .await?;
            (liked, followed)
        } else {
            (false, false)
        };

        Ok(Some(VideoDetailRow {
            id: row.id, title: row.title, description: row.description,
            file_url: row.file_url, thumbnail_url: row.thumbnail_url,
            duration: row.duration, like_count: row.like_count, total_clicks: row.total_clicks,
            author_id: row.author_id, author_name: row.author_name,
            author_avatar_url: row.author_avatar_url, author_is_vip: row.author_is_vip,
            viewer_liked, viewer_followed,
        }))
    }

    async fn delete_video(&self, video_id: i64, customer_id: i64) -> Result<bool, AppError> {
        let result = sqlx::query(
            r"UPDATE videos SET deleted_at = NOW(), updated_at = NOW()
               WHERE id = $1 AND customer_id = $2 AND status = 3 AND deleted_at IS NULL",
        )
        .bind(video_id)
        .bind(customer_id)
        .execute(self.pool)
        .await?;
        Ok(result.rows_affected() > 0)
    }

    async fn list_user_videos(&self, user_id: i64) -> Result<Vec<VideoListRow>, AppError> {
        let items = sqlx::query_as::<_, VideoListRow>(
            r"SELECT id, customer_id, title, thumbnail_url, duration, like_count, total_clicks
               FROM videos WHERE customer_id = $1 AND status = 4 AND deleted_at IS NULL
               ORDER BY id DESC",
        )
        .bind(user_id)
        .fetch_all(self.pool)
        .await?;
        Ok(items)
    }
}

/// Partial row for video detail query (without viewer interaction fields
/// that require separate queries).
#[derive(sqlx::FromRow)]
struct PgVideoDetailPartial {
    id: i64,
    title: Option<String>,
    description: Option<String>,
    file_url: String,
    thumbnail_url: Option<String>,
    duration: Option<i32>,
    like_count: i64,
    total_clicks: i64,
    author_id: i64,
    author_name: Option<String>,
    author_avatar_url: Option<String>,
    author_is_vip: bool,
}

File: kotlin/src/main/kotlin/lab/gabon/repository/VideoRepo.kt
package lab.gabon.repository

import kotlinx.datetime.Instant
import lab.gabon.config.dbQuery
import lab.gabon.model.VideoStatus
import org.jetbrains.exposed.v1.core.LikePattern
import org.jetbrains.exposed.v1.core.ResultRow
import org.jetbrains.exposed.v1.core.SortOrder
import org.jetbrains.exposed.v1.core.and
import org.jetbrains.exposed.v1.core.eq
import org.jetbrains.exposed.v1.core.isNull
import org.jetbrains.exposed.v1.core.like
import org.jetbrains.exposed.v1.core.longLiteral
import org.jetbrains.exposed.v1.core.lowerCase
import org.jetbrains.exposed.v1.core.plus
import org.jetbrains.exposed.v1.datetime.CurrentTimestampWithTimeZone
import org.jetbrains.exposed.v1.jdbc.insertAndGetId
import org.jetbrains.exposed.v1.jdbc.selectAll
import org.jetbrains.exposed.v1.jdbc.transactions.TransactionManager
import org.jetbrains.exposed.v1.jdbc.update
import java.time.OffsetDateTime

data class VideoRow(
    val id: Long,
    val customerId: Long,
    val title: String?,
    val description: String?,
    val fileName: String,
    val fileSize: Long,
    val fileUrl: String,
    val thumbnailUrl: String?,
    val mimeType: String,
    val duration: Int?,
    val status: Short,
    val totalClicks: Long,
    val validClicks: Long,
    val likeCount: Long,
    val createdAt: Instant,
    val updatedAt: Instant,
    // joined uploader info
    val uploaderName: String?,
    val uploaderAvatar: String?,
)

data class VideoListRow(
    val id: Long,
    val customerId: Long,
    val title: String?,
    val fileName: String,
    val fileUrl: String,
    val thumbnailUrl: String?,
    val mimeType: String,
    val status: Short,
    val totalClicks: Long,
    val validClicks: Long,
    val likeCount: Long,
    val createdAt: Instant,
    val uploaderName: String?,
    val uploaderAvatar: String?,
)

class VideoRepo {
    suspend fun create(
        customerId: Long,
        title: String?,
        description: String?,
        fileName: String,
        fileSize: Long,
        fileUrl: String,
        mimeType: String,
        status: Short = VideoStatus.PENDING_REVIEW.value,
    ): Long =
        dbQuery {
            Videos
                .insertAndGetId {
                    it[Videos.customerId] = customerId
                    it[Videos.title] = title
                    it[Videos.description] = description
                    it[Videos.fileName] = fileName
                    it[Videos.fileSize] = fileSize
                    it[Videos.fileUrl] = fileUrl
                    it[Videos.mimeType] = mimeType
                    it[Videos.status] = status
                }.value
        }

    suspend fun findById(id: Long): VideoRow? =
        dbQuery {
            (Videos innerJoin Customers)
                .selectAll()
                .where {
                    (Videos.id eq id) and Videos.deletedAt.isNull()
                }.singleOrNull()
                ?.toVideoRow()
        }

    suspend fun isLikedBy(
        videoId: Long,
        customerId: Long,
    ): Boolean =
        dbQuery {
            VideoLikes
                .selectAll()
                .where {
                    (VideoLikes.videoId eq videoId) and (VideoLikes.customerId eq customerId)
                }.count() > 0
        }

    suspend fun listApproved(
        page: Int,
        pageSize: Int,
        keyword: String? = null,
    ): Pair<List<VideoListRow>, Long> =
        dbQuery {
            val baseCondition = (Videos.status eq VideoStatus.APPROVED.value) and Videos.deletedAt.isNull()
            val condition =
                if (!keyword.isNullOrBlank()) {
                    baseCondition and (Videos.title.lowerCase() like LikePattern("%${keyword.lowercase()}%"))
                } else {
                    baseCondition
                }

            val total =
                (Videos innerJoin Customers)
                    .selectAll()
                    .where { condition }
                    .count()

            val items =
                (Videos innerJoin Customers)
                    .selectAll()
                    .where { condition }
                    .orderBy(Videos.createdAt to SortOrder.DESC)
                    .limit(pageSize)
                    .offset(((page - 1) * pageSize).toLong())
                    .map { it.toVideoListRow() }

            items to total
        }

    suspend fun listFeatured(
        page: Int,
        pageSize: Int,
    ): Pair<List<VideoListRow>, Long> =
        dbQuery {
            val condition = (Videos.status eq VideoStatus.APPROVED.value) and Videos.deletedAt.isNull()

            val total =
                (Videos innerJoin Customers)
                    .selectAll()
                    .where { condition }
                    .count()

            val items =
                (Videos innerJoin Customers)
                    .selectAll()
                    .where { condition }
                    .orderBy(Videos.likeCount to SortOrder.DESC)
                    .limit(pageSize)
                    .offset(((page - 1) * pageSize).toLong())
                    .map { it.toVideoListRow() }

            items to total
        }

    suspend fun listByCustomer(
        customerId: Long,
        page: Int,
        pageSize: Int,
        status: Short? = null,
    ): Pair<List<VideoListRow>, Long> =
        dbQuery {
            val baseCondition = (Videos.customerId eq customerId) and Videos.deletedAt.isNull()
            val condition =
                if (status != null) {
                    baseCondition and (Videos.status eq status)
                } else {
                    baseCondition
                }

            val total =
                (Videos innerJoin Customers)
                    .selectAll()
                    .where { condition }
                    .count()

            val items =
                (Videos innerJoin Customers)
                    .selectAll()
                    .where { condition }
                    .orderBy(Videos.createdAt to SortOrder.DESC)
                    .limit(pageSize)
                    .offset(((page - 1) * pageSize).toLong())
                    .map { it.toVideoListRow() }

            items to total
        }

    suspend fun listApprovedByUser(
        userId: Long,
        page: Int,
        pageSize: Int,
    ): Pair<List<VideoListRow>, Long> =
        dbQuery {
            val condition =
                (Videos.customerId eq userId) and
                    (Videos.status eq VideoStatus.APPROVED.value) and
                    Videos.deletedAt.isNull()

            val total =
                (Videos innerJoin Customers)
                    .selectAll()
                    .where { condition }
                    .count()

            val items =
                (Videos innerJoin Customers)
                    .selectAll()
                    .where { condition }
                    .orderBy(Videos.createdAt to SortOrder.DESC)
                    .limit(pageSize)
                    .offset(((page - 1) * pageSize).toLong())
                    .map { it.toVideoListRow() }

            items to total
        }

    suspend fun softDelete(
        id: Long,
        customerId: Long,
    ): Boolean =
        dbQuery {
            Videos.update({
                (Videos.id eq id) and (Videos.customerId eq customerId) and Videos.deletedAt.isNull()
            }) {
                it[deletedAt] = CurrentTimestampWithTimeZone
                it[updatedAt] = CurrentTimestampWithTimeZone
            } > 0
        }

    suspend fun incrementTotalClicks(id: Long): Unit =
        dbQuery {
            Videos.update({ Videos.id eq id }) {
                it[totalClicks] = totalClicks plus longLiteral(1)
            }
        }

    suspend fun incrementValidClicks(id: Long): Unit =
        dbQuery {
            Videos.update({ Videos.id eq id }) {
                it[validClicks] = validClicks plus longLiteral(1)
            }
        }

    /**
     * Atomically insert a like and increment like_count using a CTE.
     * ON CONFLICT DO NOTHING makes this idempotent — duplicate likes won't increment.
     * Returns true if the like was actually inserted (not a duplicate).
     */
    @Suppress("MagicNumber")
    suspend fun likeVideo(
        videoId: Long,
        customerId: Long,
    ): Boolean =
        dbQuery {
            val sql =
                """
                WITH inserted AS (
                    INSERT INTO video_likes (video_id, customer_id)
                    VALUES (?, ?)
                    ON CONFLICT (video_id, customer_id) DO NOTHING
                    RETURNING id
                )
                UPDATE videos SET like_count = like_count + 1, updated_at = NOW()
                WHERE id = ? AND EXISTS (SELECT 1 FROM inserted)
                """.trimIndent()

            val jdbcConn = TransactionManager.current().connection.connection as java.sql.Connection
            jdbcConn.prepareStatement(sql).use { stmt ->
                stmt.setLong(1, videoId)
                stmt.setLong(2, customerId)
                stmt.setLong(3, videoId)
                stmt.executeUpdate() > 0
            }
        }

    /**
     * Atomically delete a like and decrement like_count using a CTE.
     * If the user hasn't liked the video, DELETE returns no rows so UPDATE doesn't fire.
     * GREATEST prevents negative counts.
     * Returns true if the like was actually removed.
     */
    @Suppress("MagicNumber")
    suspend fun unlikeVideo(
        videoId: Long,
        customerId: Long,
    ): Boolean =
        dbQuery {
            val sql =
                """
                WITH deleted AS (
                    DELETE FROM video_likes
                    WHERE video_id = ? AND customer_id = ?
                    RETURNING id
                )
                UPDATE videos SET like_count = GREATEST(like_count - 1, 0), updated_at = NOW()
                WHERE id = ? AND EXISTS (SELECT 1 FROM deleted)
                """.trimIndent()

            val jdbcConn = TransactionManager.current().connection.connection as java.sql.Connection
            jdbcConn.prepareStatement(sql).use { stmt ->
                stmt.setLong(1, videoId)
                stmt.setLong(2, customerId)
                stmt.setLong(3, videoId)
                stmt.executeUpdate() > 0
            }
        }

    private fun ResultRow.toVideoRow(): VideoRow =
        VideoRow(
            id = this[Videos.id].value,
            customerId = this[Videos.customerId].value,
            title = this[Videos.title],
            description = this[Videos.description],
            fileName = this[Videos.fileName],
            fileSize = this[Videos.fileSize],
            fileUrl = this[Videos.fileUrl],
            thumbnailUrl = this[Videos.thumbnailUrl],
            mimeType = this[Videos.mimeType],
            duration = this[Videos.duration],
            status = this[Videos.status],
            totalClicks = this[Videos.totalClicks],
            validClicks = this[Videos.validClicks],
            likeCount = this[Videos.likeCount],
            createdAt = this[Videos.createdAt].toKotlinInstant(),
            updatedAt = this[Videos.updatedAt].toKotlinInstant(),
            uploaderName = this[Customers.name],
            uploaderAvatar = this[Customers.avatarUrl],
        )

    private fun ResultRow.toVideoListRow(): VideoListRow =
        VideoListRow(
            id = this[Videos.id].value,
            customerId = this[Videos.customerId].value,
            title = this[Videos.title],
            fileName = this[Videos.fileName],
            fileUrl = this[Videos.fileUrl],
            thumbnailUrl = this[Videos.thumbnailUrl],
            mimeType = this[Videos.mimeType],
            status = this[Videos.status],
            totalClicks = this[Videos.totalClicks],
            validClicks = this[Videos.validClicks],
            likeCount = this[Videos.likeCount],
            createdAt = this[Videos.createdAt].toKotlinInstant(),
            uploaderName = this[Customers.name],
            uploaderAvatar = this[Customers.avatarUrl],
        )

    private fun OffsetDateTime.toKotlinInstant(): Instant = Instant.fromEpochSeconds(toEpochSecond(), nano)
}
